File: desktop/native/SurveyController.App/Views/StrategyEditor.Weights.cpp
#include "pch.h"
#include "StrategyEditor.xaml.h"
#include "ViewModels/OptionWeight.h"
#include "Services/JsonHelpers.h"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <sstream>

namespace winrt::SurveyController::App::implementation
{
    namespace
    {
        using namespace Windows::Data::Json;

        JsonObject ConfiguredWeights(JsonObject const& strategy)
        {
            auto table = Services::GetJsonObject(strategy, L"custom_weights");
            if (Services::GetJsonArray(table, L"options").Size() > 0 ||
                Services::GetJsonArray(table, L"rows").Size() > 0) return table;
            return Services::GetJsonObject(strategy, L"probabilities");
        }

        hstring PercentText(double value)
        {
            std::wostringstream output;
            if (std::abs(value - std::round(value)) < 0.05)
            {
                output << static_cast<int32_t>(std::round(value));
            }
            else
            {
                output << std::fixed << std::setprecision(1) << value;
            }
            output << L'%';
            return hstring{ output.str() };
        }

        double SliderNumber(hstring const& value, double fallback)
        {
            try
            {
                size_t parsed{};
                auto text = std::wstring{ value };
                auto result = std::stod(text, &parsed);
                return parsed == text.size() && std::isfinite(result) ? result : fallback;
            }
            catch (...) { return fallback; }
        }

        hstring NumberText(double value)
        {
            std::wostringstream output;
            if (std::abs(value - std::round(value)) < 0.0001) output << static_cast<int64_t>(std::llround(value));
            else output << std::fixed << std::setprecision(2) << value;
            return hstring{ output.str() };
        }
    }

    void StrategyEditor::RebuildWeightEditor(JsonObject const& question, JsonObject const& strategy,
        hstring const& normalizedType)
    {
        m_syncingWeights = true;
        m_weightOptions.Clear();
        m_weightLabels.clear();

        auto isTextQuestion = normalizedType == L"text" || normalizedType == L"multi_text" || normalizedType == L"location";
        auto optionTexts = Services::GetJsonArray(question, L"option_texts");
        auto weightTable = ConfiguredWeights(strategy);
        auto optionWeights = Services::GetJsonArray(weightTable, L"options");
        auto rowWeights = Services::GetJsonArray(weightTable, L"rows");
        auto optionCount = static_cast<uint32_t>((std::max)(0.0,
            strategy.GetNamedNumber(L"option_count", question.GetNamedNumber(L"options", 0))));
        optionCount = (std::max)(optionCount, optionTexts.Size());
        optionCount = (std::max)(optionCount, optionWeights.Size());
        for (auto const& row : rowWeights)
        {
            if (row.ValueType() == JsonValueType::Array) optionCount = (std::max)(optionCount, row.GetArray().Size());
        }
        if (isTextQuestion || normalizedType == L"unsupported") optionCount = 0;
        if (normalizedType == L"slider") optionCount = 1;

        auto rowTexts = Services::GetJsonArray(question, L"row_texts");
        auto rowCount = static_cast<uint32_t>((std::max)(0.0,
            strategy.GetNamedNumber(L"rows", question.GetNamedNumber(L"rows", 0))));
        rowCount = (std::max)(rowCount, rowTexts.Size());
        rowCount = (std::max)(rowCount, rowWeights.Size());
        auto matrix = normalizedType == L"matrix";
        m_weightRows = matrix && rowCount > 0 ? rowCount : 1;
        m_weightColumns = optionCount;

        m_multipleWeights = normalizedType == L"multiple";
        m_sliderValue = normalizedType == L"slider";
        Bias().Visibility(m_sliderValue ? Microsoft::UI::Xaml::Visibility::Collapsed : Microsoft::UI::Xaml::Visibility::Visible);
        OptionWeightsSection().Visibility(optionCount > 0
            ? Microsoft::UI::Xaml::Visibility::Visible
            : Microsoft::UI::Xaml::Visibility::Collapsed);
        OptionWeightEmpty().Visibility(optionCount == 0
            ? Microsoft::UI::Xaml::Visibility::Visible
            : Microsoft::UI::Xaml::Visibility::Collapsed);

        auto minimum = m_sliderValue ? SliderNumber(question.GetNamedString(L"slider_min", L"0"), 0) : 0.0;
        auto maximum = m_sliderValue ? SliderNumber(question.GetNamedString(L"slider_max", L"100"), 100)
            : (m_multipleWeights ? 100.0 : 50.0);
        if (maximum < minimum) std::swap(minimum, maximum);
        auto step = m_sliderValue ? SliderNumber(question.GetNamedString(L"slider_step", L"1"), 1) : 1.0;
        if (step <= 0) step = 1;
        auto defaultValue = m_sliderValue ? (minimum + maximum) / 2.0 : (m_multipleWeights ? 50.0 : 1.0);
        for (uint32_t rowIndex = 0; rowIndex < m_weightRows; ++rowIndex)
        {
            auto configuredRow = rowIndex < rowWeights.Size() && rowWeights.GetAt(rowIndex).ValueType() == JsonValueType::Array
                ? rowWeights.GetArrayAt(rowIndex) : JsonArray{};
            auto rowLabel = rowIndex < rowTexts.Size() ? rowTexts.GetStringAt(rowIndex)
                : hstring{ L"矩阵行 " + std::to_wstring(rowIndex + 1) };
            for (uint32_t optionIndex = 0; optionIndex < optionCount; ++optionIndex)
            {
                auto optionLabel = m_sliderValue ? hstring{ L"滑块值" } : optionIndex < optionTexts.Size()
                    ? optionTexts.GetStringAt(optionIndex)
                    : hstring{ L"选项 " + std::to_wstring(optionIndex + 1) };
                if (optionLabel.empty()) optionLabel = hstring{ L"选项 " + std::to_wstring(optionIndex + 1) };
                auto label = m_weightRows > 1
                    ? hstring{ std::wstring{ rowLabel } + L" · " + std::wstring{ optionLabel } }
                    : optionLabel;
                m_weightLabels.push_back(label);
                auto values = m_weightRows > 1 ? configuredRow : optionWeights;
                auto value = optionIndex < values.Size() && values.GetAt(optionIndex).ValueType() == JsonValueType::Number
                    ? values.GetNumberAt(optionIndex) : defaultValue;
                value = (std::max)(minimum, (std::min)(maximum, value));
                value = minimum + std::round((value - minimum) / step) * step;
                value = (std::max)(minimum, (std::min)(maximum, value));
                auto option = winrt::make<implementation::OptionWeight>(label, value, minimum, maximum, step);
                auto weak = get_weak();
                option.PropertyChanged([weak](auto const&, auto const&)
                {
                    if (auto self = weak.get(); self && !self->m_syncingWeights)
                    {
                        self->m_currentQuestionDirty = true;
                        self->SelectTag(self->Bias(), L"custom");
                        self->UpdateRatioPreview();
                    }
                });
                m_weightOptions.Append(option);
            }
        }
        m_syncingWeights = false;
        UpdateRatioPreview();
    }

    Windows::Foundation::Collections::IObservableVector<SurveyController::App::OptionWeight>
        StrategyEditor::WeightOptions()
    {
        return m_weightOptions;
    }

    void StrategyEditor::OnBiasChanged(IInspectable const&,
        Microsoft::UI::Xaml::Controls::SelectionChangedEventArgs const&)
    {
        if (!m_initialized || m_syncingWeights) return;
        m_currentQuestionDirty = true;
        auto bias = SelectedTag(Bias(), L"custom");
        if (bias != L"custom") ApplyBiasPreset(bias);
    }

    void StrategyEditor::ApplyBiasPreset(hstring const& bias)
    {
        if (m_sliderValue) return;
        auto columns = m_weightColumns;
        if (columns == 0) return;
        std::vector<double> raw(columns, 1.0);
        if (columns > 1)
        {
            auto center = static_cast<double>(columns - 1) / 2.0;
            for (size_t index = 0; index < columns; ++index)
            {
                auto position = static_cast<double>(index) / static_cast<double>(columns - 1);
                auto base = bias == L"left" ? 1.0 - position : position;
                if (bias == L"center") base = 1.0 - std::abs(static_cast<double>(index) - center) / center;
                raw[index] = std::pow((std::max)(0.0, base), bias == L"center" ? 3.0 : 8.0);
            }
        }
        auto maximum = *std::max_element(raw.begin(), raw.end());
        auto scale = m_multipleWeights ? 100.0 : 50.0;
        for (uint32_t row = 0; row < m_weightRows; ++row)
        {
            for (uint32_t index = 0; index < columns; ++index)
            {
                auto value = maximum > 0 ? std::round(raw[index] / maximum * scale) : std::round(scale / columns);
                m_weightOptions.GetAt(row * columns + index).Value(value);
            }
        }
    }

    Windows::Data::Json::JsonArray StrategyEditor::WeightValues() const
    {
        Windows::Data::Json::JsonArray values;
        for (auto const& option : m_weightOptions)
        {
            auto value = option.Value();
            values.Append(Windows::Data::Json::JsonValue::CreateNumberValue(std::isnan(value) ? 0 : value));
        }
        return values;
    }

    Windows::Data::Json::JsonObject StrategyEditor::WeightTable() const
    {
        Windows::Data::Json::JsonObject table;
        auto values = WeightValues();
        if (m_weightRows <= 1)
        {
            table.SetNamedValue(L"options", values);
            return table;
        }
        Windows::Data::Json::JsonArray rows;
        for (uint32_t row = 0; row < m_weightRows; ++row)
        {
            Windows::Data::Json::JsonArray rowValues;
            for (uint32_t column = 0; column < m_weightColumns; ++column)
            {
                rowValues.Append(values.GetAt(row * m_weightColumns + column));
            }
            rows.Append(rowValues);
        }
        table.SetNamedValue(L"rows", rows);
        return table;
    }

    void StrategyEditor::UpdateRatioPreview()
    {
        if (m_weightOptions.Size() == 0)
        {
            RatioPreview().Text(L"");
            return;
        }
        if (m_sliderValue)
        {
            RatioPreview().Text(hstring{ L"预计值：" + std::wstring{ NumberText(m_weightOptions.GetAt(0).Value()) } });
            return;
        }
        std::vector<double> values;
        values.reserve(m_weightOptions.Size());
        for (auto const& option : m_weightOptions)
        {
            auto value = std::isnan(option.Value()) ? 0 : (std::max)(0.0, option.Value());
            values.push_back(value);
        }

        std::wstring preview;
        auto columns = (std::max)(1u, m_weightColumns);
        for (uint32_t row = 0; row < m_weightRows; ++row)
        {
            if (row > 0) preview += L"\n";
            preview += m_multipleWeights ? L"命中率：" : L"预计占比：";
            double rowTotal = 0;
            for (uint32_t column = 0; column < columns; ++column) rowTotal += values[row * columns + column];
            for (uint32_t column = 0; column < columns; ++column)
            {
                auto index = row * columns + column;
                if (column > 0) preview += L" | ";
                std::wstring label{ m_weightLabels[index] };
                if (label.size() > 14) label = label.substr(0, 14) + L"…";
                auto percent = m_multipleWeights ? values[index]
                    : (rowTotal > 0 ? values[index] / rowTotal * 100.0 : 100.0 / columns);
                preview += label + L" " + std::wstring{ PercentText(percent) };
            }
        }
        RatioPreview().Text(hstring{ preview });
    }
}

File: desktop/native-cs/src/SurveyController.App/Views/StrategyEditor.Weights.cs
using System.Globalization;
using System.Text.Json.Nodes;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using SurveyController.App.ViewModels;

namespace SurveyController.App.Views;

/// <summary>权重表构建、倾向预设与占比预览，对照 C++ StrategyEditor.Weights.cpp。</summary>
public sealed partial class StrategyEditor
{
    private static JsonObject? ConfiguredWeights(JsonObject strategy)
    {
        var table = strategy["custom_weights"] as JsonObject;
        var options = table?["options"] as JsonArray;
        var rows = table?["rows"] as JsonArray;
        if ((options?.Count ?? 0) > 0 || (rows?.Count ?? 0) > 0)
        {
            return table;
        }
        return strategy["probabilities"] as JsonObject;
    }

    internal static string PercentText(double value)
    {
        if (Math.Abs(value - Math.Round(value)) < 0.05)
        {
            return $"{(int)Math.Round(value)}%";
        }
        return value.ToString("0.0", CultureInfo.InvariantCulture) + "%";
    }

    private static double SliderNumber(string value, double fallback) =>
        double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed)
            && double.IsFinite(parsed) ? parsed : fallback;

    internal static string NumberText(double value) =>
        Math.Abs(value - Math.Round(value)) < 0.0001
            ? ((long)Math.Round(value)).ToString(CultureInfo.InvariantCulture)
            : value.ToString("0.00", CultureInfo.InvariantCulture);

    protected void RebuildWeightEditor(JsonObject question, JsonObject strategy, string normalizedType)
    {
        SyncingWeights = true;
        WeightOptions.Clear();
        WeightLabels.Clear();

        var isTextQuestion = normalizedType is "text" or "multi_text" or "location";
        var optionTexts = question["option_texts"] as JsonArray ?? [];
        var weightTable = ConfiguredWeights(strategy);
        var optionWeights = weightTable?["options"] as JsonArray ?? [];
        var rowWeights = weightTable?["rows"] as JsonArray ?? [];
        var optionCount = Math.Max(0, (int)JsonNumber(strategy, "option_count", JsonNumber(question, "options", 0)));
        optionCount = Math.Max(optionCount, optionTexts.Count);
        optionCount = Math.Max(optionCount, optionWeights.Count);
        foreach (var row in rowWeights)
        {
            if (row is JsonArray rowArray)
            {
                optionCount = Math.Max(optionCount, rowArray.Count);
            }
        }
        if (isTextQuestion || normalizedType == "unsupported")
        {
            optionCount = 0;
        }
        if (normalizedType == "slider")
        {
            optionCount = 1;
        }

        var rowTexts = question["row_texts"] as JsonArray ?? [];
        var rowCount = Math.Max(0, (int)JsonNumber(strategy, "rows", JsonNumber(question, "rows", 0)));
        rowCount = Math.Max(rowCount, rowTexts.Count);
        rowCount = Math.Max(rowCount, rowWeights.Count);
        var matrix = normalizedType == "matrix";
        WeightRows = matrix && rowCount > 0 ? rowCount : 1;
        WeightColumns = optionCount;

        MultipleWeights = normalizedType == "multiple";
        SliderValue = normalizedType == "slider";
        Bias.Visibility = SliderValue ? Visibility.Collapsed : Visibility.Visible;
        OptionWeightsSection.Visibility = optionCount > 0 ? Visibility.Visible : Visibility.Collapsed;
        OptionWeightEmpty.Visibility = optionCount == 0 ? Visibility.Visible : Visibility.Collapsed;

        var minimum = SliderValue
            ? SliderNumber(JsonStr(question, "slider_min", "0"), 0)
            : 0.0;
        var maximum = SliderValue
            ? SliderNumber(JsonStr(question, "slider_max", "100"), 100)
            : MultipleWeights ? 100.0 : 50.0;
        if (maximum < minimum)
        {
            (minimum, maximum) = (maximum, minimum);
        }
        var step = SliderValue ? SliderNumber(JsonStr(question, "slider_step", "1"), 1) : 1.0;
        if (step <= 0)
        {
            step = 1;
        }
        var defaultValue = SliderValue ? (minimum + maximum) / 2.0 : MultipleWeights ? 50.0 : 1.0;
        for (var rowIndex = 0; rowIndex < WeightRows; rowIndex++)
        {
            var configuredRow = rowIndex < rowWeights.Count && rowWeights[rowIndex] is JsonArray array
                ? array
                : null;
            var rowLabel = rowIndex < rowTexts.Count
                ? AsString(rowTexts[rowIndex])
                : $"矩阵行 {rowIndex + 1}";
            for (var optionIndex = 0; optionIndex < optionCount; optionIndex++)
            {
                var optionLabel = SliderValue
                    ? "滑块值"
                    : optionIndex < optionTexts.Count
                        ? AsString(optionTexts[optionIndex])
                        : $"选项 {optionIndex + 1}";
                if (optionLabel.Length == 0)
                {
                    optionLabel = $"选项 {optionIndex + 1}";
                }
                var label = WeightRows > 1 ? $"{rowLabel} · {optionLabel}" : optionLabel;
                WeightLabels.Add(label);
                var values = WeightRows > 1 ? configuredRow : optionWeights;
                var value = values is not null && optionIndex < values.Count && values[optionIndex] is System.Text.Json.Nodes.JsonValue
                    ? AsDouble(values[optionIndex])
                    : defaultValue;
                value = Math.Max(minimum, Math.Min(maximum, value));
                value = minimum + Math.Round((value - minimum) / step) * step;
                value = Math.Max(minimum, Math.Min(maximum, value));
                var option = new OptionWeight(label, value, minimum, maximum, step);
                option.PropertyChanged += (_, _) =>
                {
                    if (!SyncingWeights)
                    {
                        CurrentQuestionDirty = true;
                        SelectTag(Bias, "custom");
                        UpdateRatioPreview();
                    }
                };
                WeightOptions.Add(option);
            }
        }
        SyncingWeights = false;
        UpdateRatioPreview();
    }

    private void OnBiasChanged(object sender, SelectionChangedEventArgs e)
    {
        if (!_isLoaded || SyncingWeights)
        {
            return;
        }
        CurrentQuestionDirty = true;
        var bias = SelectedTag(Bias, "custom");
        if (bias != "custom")
        {
            ApplyBiasPreset(bias);
        }
    }

    protected void ApplyBiasPreset(string bias)
    {
        if (SliderValue)
        {
            return;
        }
        var columns = WeightColumns;
        if (columns == 0)
        {
            return;
        }
        var raw = new double[columns];
        Array.Fill(raw, 1.0);
        if (columns > 1)
        {
            var center = (columns - 1) / 2.0;
            for (var index = 0; index < columns; index++)
            {
                var position = index / (double)(columns - 1);
                var baseValue = bias == "left" ? 1.0 - position : position;
                if (bias == "center")
                {
                    baseValue = 1.0 - Math.Abs(index - center) / center;
                }
                raw[index] = Math.Pow(Math.Max(0.0, baseValue), bias == "center" ? 3.0 : 8.0);
            }
        }
        var maximum = raw.Length > 0 ? raw.Max() : 0;
        var scale = MultipleWeights ? 100.0 : 50.0;
        for (var row = 0; row < WeightRows; row++)
        {
            for (var index = 0; index < columns; index++)
            {
                var value = maximum > 0 ? Math.Round(raw[index] / maximum * scale) : Math.Round(scale / columns);
                WeightOptions[row * columns + index].Value = value;
            }
        }
    }

    private JsonArray BuildWeightValues()
    {
        var values = new JsonArray();
        foreach (var option in WeightOptions)
        {
            var value = option.Value;
            values.Add(double.IsNaN(value) ? 0 : value);
        }
        return values;
    }

    private JsonObject BuildWeightTable()
    {
        var table = new JsonObject();
        var values = BuildWeightValues();
        if (WeightRows <= 1)
        {
            table["options"] = values;
            return table;
        }
        var rows = new JsonArray();
        for (var row = 0; row < WeightRows; row++)
        {
            var rowValues = new JsonArray();
            for (var column = 0; column < WeightColumns; column++)
            {
                rowValues.Add((JsonNode?)values[row * WeightColumns + column]);
            }
            rows.Add(rowValues);
        }
        table["rows"] = rows;
        return table;
    }

    protected void UpdateRatioPreview()
    {
        if (WeightOptions.Count == 0)
        {
            RatioPreview.Text = "";
            return;
        }
        if (SliderValue)
        {
            RatioPreview.Text = "预计值：" + NumberText(WeightOptions[0].Value);
            return;
        }
        var values = new List<double>(WeightOptions.Count);
        foreach (var option in WeightOptions)
        {
            values.Add(double.IsNaN(option.Value) ? 0 : Math.Max(0.0, option.Value));
        }

        var preview = string.Empty;
        var columns = Math.Max(1, WeightColumns);
        for (var row = 0; row < WeightRows; row++)
        {
            if (row > 0)
            {
                preview += "\n";
            }
            preview += MultipleWeights ? "命中率：" : "预计占比：";
            var rowTotal = 0.0;
            for (var column = 0; column < columns; column++)
            {
                rowTotal += values[row * columns + column];
            }
            for (var column = 0; column < columns; column++)
            {
                var index = row * columns + column;
                if (column > 0)
                {
                    preview += " | ";
                }
                var label = WeightLabels[index];
                if (label.Length > 14)
                {
                    label = label[..14] + "…";
                }
                var percent = MultipleWeights
                    ? values[index]
                    : rowTotal > 0 ? values[index] / rowTotal * 100.0 : 100.0 / columns;
                preview += $"{label} {PercentText(percent)}";
            }
        }
        RatioPreview.Text = preview;
    }
}
